File: backend-ai/app/position_engine.py
"""
Position estimation — EKF GPS+IMU + map matching (spec §6).
GPS ~1 Hz + IMU predict ~50 Hz → helmet pose mượt; snap vào site polygon.
"""
from __future__ import annotations

import math
import time
from dataclasses import dataclass, field

import numpy as np

from .patrol_site_geometry import PATROL_SITE_CENTER, snap_point_to_site

M_PER_DEG_LAT = 111_320.0
GPS_DEFAULT_ACCURACY_M = 8.0
MAX_PREDICT_DT_S = 0.5


def latlon_to_enu(lat: float, lon: float, ref_lat: float, ref_lon: float) -> tuple[float, float]:
    cos_lat = math.cos(math.radians(ref_lat))
    east = (lon - ref_lon) * M_PER_DEG_LAT * cos_lat
    north = (lat - ref_lat) * M_PER_DEG_LAT
    return east, north


def enu_to_latlon(east: float, north: float, ref_lat: float, ref_lon: float) -> tuple[float, float]:
    cos_lat = math.cos(math.radians(ref_lat))
    lat = ref_lat + north / M_PER_DEG_LAT
    lon = ref_lon + east / (M_PER_DEG_LAT * max(cos_lat, 1e-6))
    return lat, lon


@dataclass
class HelmetEkf:
    """4-state ENU Kalman: east, north, v_east, v_north (m, m/s)."""

    ref_lat: float
    ref_lon: float
    x: np.ndarray = field(default_factory=lambda: np.zeros(4))
    P: np.ndarray = field(default_factory=lambda: np.eye(4) * 25.0)
    heading_deg: float = 0.0
    last_ts: float = 0.0
    initialized: bool = False

    def predict(self, dt: float, heading_deg: float | None = None) -> None:
        if dt <= 0:
            return
        dt = min(dt, MAX_PREDICT_DT_S)
        F = np.eye(4)
        F[0, 2] = dt
        F[1, 3] = dt
        q_pos = 0.08 * dt
        q_vel = 0.35 * dt
        Q = np.diag([q_pos, q_pos, q_vel, q_vel])
        self.x = F @ self.x
        self.P = F @ self.P @ F.T + Q
        if heading_deg is not None:
            self.heading_deg = float(heading_deg) % 360.0
            speed = float(math.hypot(self.x[2], self.x[3]))
            if speed < 0.45:
                br = math.radians(self.heading_deg)
                self.x[2] = 0.25 * math.sin(br)
                self.x[3] = 0.25 * math.cos(br)

    def update_gps(self, lat: float, lon: float, accuracy_m: float = GPS_DEFAULT_ACCURACY_M) -> None:
        east, north = latlon_to_enu(lat, lon, self.ref_lat, self.ref_lon)
        if not self.initialized:
            self.x[0], self.x[1] = east, north
            self.x[2] = self.x[3] = 0.0
            self.initialized = True
            var = max(4.0, float(accuracy_m) ** 2)
            self.P = np.eye(4) * var
            return
        z = np.array([east, north])
        H = np.array([[1.0, 0.0, 0.0, 0.0], [0.0, 1.0, 0.0, 0.0]])
        r_var = max(4.0, float(accuracy_m) ** 2)
        R = np.eye(2) * r_var
        y = z - H @ self.x
        S = H @ self.P @ H.T + R
        K = self.P @ H.T @ np.linalg.inv(S)
        self.x = self.x + K @ y
        self.P = (np.eye(4) - K @ H) @ self.P

    def lat_lon(self) -> tuple[float, float] | None:
        if not self.initialized:
            return None
        return enu_to_latlon(float(self.x[0]), float(self.x[1]), self.ref_lat, self.ref_lon)


class HelmetPositionTracker:
    """Per-helmet EKF + map matching."""

    def __init__(self) -> None:
        self._ekf: dict[str, HelmetEkf] = {}

    def _get(self, helmet_id: str) -> HelmetEkf:
        if helmet_id not in self._ekf:
            ref_lat, ref_lon = PATROL_SITE_CENTER
            self._ekf[helmet_id] = HelmetEkf(ref_lat=ref_lat, ref_lon=ref_lon)
        return self._ekf[helmet_id]

    def fuse_helmet_pose(
        self,
        helmet_id: str,
        *,
        lat: float | None = None,
        lon: float | None = None,
        heading: float | None = None,
        pitch: float | None = None,
        roll: float | None = None,
        accuracy_m: float | None = None,
        ts: float | None = None,
    ) -> tuple[float | None, float | None, float | None, str]:
        """
        Returns (lat, lon, heading, method) after EKF + map match.
        method: raw | ekf | ekf_map | imu_only
        """
        del pitch, roll  # reserved for future 6-DOF; heading drives predict
        now = ts if ts is not None else time.time()
        ekf = self._get(helmet_id)
        if ekf.last_ts > 0:
            ekf.predict(now - ekf.last_ts, heading_deg=heading)
        ekf.last_ts = now

        method = "imu_only"
        out_lat: float | None = None
        out_lon: float | None = None
        out_heading = (float(heading) % 360.0) if heading is not None else ekf.heading_deg

        has_gps = (
            lat is not None
            and lon is not None
            and math.isfinite(lat)
            and math.isfinite(lon)
            and not (lat == 0.0 and lon == 0.0)
        )
        if has_gps:
            ekf.update_gps(float(lat), float(lon), accuracy_m or GPS_DEFAULT_ACCURACY_M)
            method = "ekf"
            pair = ekf.lat_lon()
            if pair:
                out_lat, out_lon = pair

        if out_lat is None and ekf.initialized:
            pair = ekf.lat_lon()
            if pair:
                out_lat, out_lon = pair

        if out_lat is not None and out_lon is not None:
            matched_lat, matched_lon, _inside = snap_point_to_site(out_lat, out_lon)
            out_lat, out_lon = matched_lat, matched_lon
            if method == "ekf":
                method = "ekf_map"

        if heading is not None:
            ekf.heading_deg = float(heading) % 360.0
            out_heading = ekf.heading_deg

        return out_lat, out_lon, out_heading, method

    def map_match_object(self, lat: float, lon: float) -> tuple[float, float]:
        matched_lat, matched_lon, _ = snap_point_to_site(lat, lon)
        return matched_lat, matched_lon

    def reset(self, helmet_id: str | None = None) -> None:
        if helmet_id is None:
            self._ekf.clear()
        else:
            self._ekf.pop(helmet_id, None)


helmet_position_tracker = HelmetPositionTracker()


def fuse_helmet_pose(
    helmet_id: str,
    *,
    lat: float | None = None,
    lon: float | None = None,
    heading: float | None = None,
    pitch: float | None = None,
    roll: float | None = None,
    accuracy_m: float | None = None,
    ts: float | None = None,
) -> tuple[float | None, float | None, float | None, str]:
    return helmet_position_tracker.fuse_helmet_pose(
        helmet_id,
        lat=lat,
        lon=lon,
        heading=heading,
        pitch=pitch,
        roll=roll,
        accuracy_m=accuracy_m,
        ts=ts,
    )


def map_match_position(lat: float, lon: float) -> tuple[float, float]:
    return helmet_position_tracker.map_match_object(lat, lon)

File: src/modules/module05-productivity/utils/positionEngine.ts
/**
 * Position engine — EKF GPS+IMU + map matching (spec §6).
 * FE mirror of backend-ai/app/position_engine.py
 */
import {
  clampPointToSiteBoundary,
  isPointInSiteBoundary,
  PATROL_SITE_CORNERS,
} from '../data/patrolSiteGeometry'
import { PATROL_SITE_CENTER } from '../data/patrolSiteMap'
import { isPatrolRelativeGpsCamera } from './patrolGpsConfig'

const M_PER_DEG_LAT = 111_320
const GPS_DEFAULT_ACCURACY_M = 8
const MAX_PREDICT_DT_S = 0.5
/** Giới hạn drift tương đối (~120m) — tránh GPS nhảy xa khỏi công trường. */
const MAX_RELATIVE_OFFSET_M = 120

export type HelmetPositionMethod =
  | 'raw'
  | 'ekf'
  | 'ekf_map'
  | 'map'
  | 'imu_only'
  | 'relative'
  | 'relative_ekf'

function latLonToEnu(
  lat: number,
  lng: number,
  refLat: number,
  refLng: number,
): [number, number] {
  const cosLat = Math.cos((refLat * Math.PI) / 180)
  const east = (lng - refLng) * M_PER_DEG_LAT * cosLat
  const north = (lat - refLat) * M_PER_DEG_LAT
  return [east, north]
}

function enuToLatLon(
  east: number,
  north: number,
  refLat: number,
  refLng: number,
): [number, number] {
  const cosLat = Math.cos((refLat * Math.PI) / 180)
  const lat = refLat + north / M_PER_DEG_LAT
  const lng = refLng + east / (M_PER_DEG_LAT * Math.max(cosLat, 1e-6))
  return [lat, lng]
}

function isValidGps(lat: number, lng: number): boolean {
  return Number.isFinite(lat)
    && Number.isFinite(lng)
    && !(lat === 0 && lng === 0)
}

function nearestOnSegment(
  px: number,
  py: number,
  ax: number,
  ay: number,
  bx: number,
  by: number,
): [number, number, number] {
  const abx = bx - ax
  const aby = by - ay
  const apx = px - ax
  const apy = py - ay
  const denom = abx * abx + aby * aby
  if (denom <= 1e-18) {
    return [ax, ay, Math.hypot(px - ax, py - ay)]
  }
  const t = Math.max(0, Math.min(1, (apx * abx + apy * aby) / denom))
  const qx = ax + t * abx
  const qy = ay + t * aby
  return [qx, qy, Math.hypot(px - qx, py - qy)]
}

/** Map matching — snap ngoài ranh lên cạnh polygon site. */
export function snapPointToSite(lat: number, lng: number): [number, number, boolean] {
  if (isPointInSiteBoundary(lat, lng)) {
    return [lat, lng, true]
  }
  let bestLat = lat
  let bestLng = lng
  let bestD = Infinity
  const px = lng
  const py = lat
  const ring = PATROL_SITE_CORNERS
  for (let i = 0; i < ring.length; i += 1) {
    const [aLat, aLng] = ring[i]
    const [bLat, bLng] = ring[(i + 1) % ring.length]
    const [qx, qy, d] = nearestOnSegment(px, py, aLng, aLat, bLng, bLat)
    if (d < bestD) {
      bestD = d
      bestLat = qy
      bestLng = qx
    }
  }
  if (!isPointInSiteBoundary(bestLat, bestLng)) {
    return [...clampPointToSiteBoundary(lat, lng), false] as [number, number, boolean]
  }
  return [
    parseFloat(bestLat.toFixed(6)),
    parseFloat(bestLng.toFixed(6)),
    false,
  ]
}

class HelmetEkf {
  private refLat: number
  private refLng: number
  private x = [0, 0, 0, 0]
  private p = [25, 0, 0, 0, 0, 25, 0, 0, 0, 0, 25, 0, 0, 0, 0, 25]
  headingDeg = 0
  lastTs = 0
  initialized = false

  constructor(refLat: number, refLng: number) {
    this.refLat = refLat
    this.refLng = refLng
  }

  predict(dt: number, headingDeg?: number | null): void {
    if (dt <= 0) return
    const step = Math.min(dt, MAX_PREDICT_DT_S)
    this.x[0] += this.x[2] * step
    this.x[1] += this.x[3] * step
    const q = 0.15 * step
    this.p[0] += q
    this.p[5] += q
    this.p[10] += q * 2
    this.p[15] += q * 2
    if (headingDeg != null && Number.isFinite(headingDeg)) {
      this.headingDeg = ((headingDeg % 360) + 360) % 360
      const speed = Math.hypot(this.x[2], this.x[3])
      if (speed < 0.45) {
        const br = (this.headingDeg * Math.PI) / 180
        this.x[2] = 0.25 * Math.sin(br)
        this.x[3] = 0.25 * Math.cos(br)
      }
    }
  }

  updateGps(lat: number, lng: number, accuracyM = GPS_DEFAULT_ACCURACY_M): void {
    const [east, north] = latLonToEnu(lat, lng, this.refLat, this.refLng)
    if (!this.initialized) {
      this.x[0] = east
      this.x[1] = north
      this.x[2] = 0
      this.x[3] = 0
      this.initialized = true
      const v = Math.max(4, accuracyM ** 2)
      this.p = [v, 0, 0, 0, 0, v, 0, 0, 0, 0, v, 0, 0, 0, 0, v]
      return
    }
    const r = Math.max(4, accuracyM ** 2)
    const k0 = this.p[0] / (this.p[0] + r)
    const k1 = this.p[5] / (this.p[5] + r)
    this.x[0] += k0 * (east - this.x[0])
    this.x[1] += k1 * (north - this.x[1])
    this.p[0] *= (1 - k0)
    this.p[5] *= (1 - k1)
  }

  latLng(): [number, number] | null {
    if (!this.initialized) return null
    return enuToLatLon(this.x[0], this.x[1], this.refLat, this.refLng)
  }
}

const ekfByHelmet = new Map<string, HelmetEkf>()
/** Mốc GPS thật lần đầu — delta cộng vào PATROL_SITE_CENTER. */
const gpsAnchorByHelmet = new Map<string, { lat: number; lng: number }>()

function clampRelativeOffsetMeters(eastM: number, northM: number): [number, number] {
  const dist = Math.hypot(eastM, northM)
  if (dist <= MAX_RELATIVE_OFFSET_M || dist <= 1e-6) return [eastM, northM]
  const scale = MAX_RELATIVE_OFFSET_M / dist
  return [eastM * scale, northM * scale]
}

/**
 * Demo GPS: vị trí hiển thị = site center + (GPS hiện tại − GPS mốc ban đầu).
 * Lần fix đầu → neo tại center; đi bộ ở Hà Nội vẫn thấy di chuyển trong Cầu Sông Hốt.
 */
export function mapRelativeGpsToSite(
  cameraId: string,
  lat: number,
  lng: number,
): [number, number] {
  const anchor = gpsAnchorByHelmet.get(cameraId)
  if (!anchor) {
    gpsAnchorByHelmet.set(cameraId, { lat, lng })
    return [PATROL_SITE_CENTER[0], PATROL_SITE_CENTER[1]]
  }

  const [dEast, dNorth] = latLonToEnu(lat, lng, anchor.lat, anchor.lng)
  const [cEast, cNorth] = clampRelativeOffsetMeters(dEast, dNorth)
  return enuToLatLon(
    cEast,
    cNorth,
    PATROL_SITE_CENTER[0],
    PATROL_SITE_CENTER[1],
  )
}

function getEkf(helmetId: string): HelmetEkf {
  let ekf = ekfByHelmet.get(helmetId)
  if (!ekf) {
    ekf = new HelmetEkf(PATROL_SITE_CENTER[0], PATROL_SITE_CENTER[1])
    ekfByHelmet.set(helmetId, ekf)
  }
  return ekf
}

export interface FuseHelmetPoseInput {
  cameraId: string
  lat?: number | null
  lng?: number | null
  heading?: number | null
  accuracyM?: number
  ts?: number
}

export interface FusedHelmetPose {
  lat: number | null
  lng: number | null
  heading: number | null
  method: HelmetPositionMethod
}

export function fuseHelmetPose(input: FuseHelmetPoseInput): FusedHelmetPose {
  const now = input.ts ?? Date.now()
  const ekf = getEkf(input.cameraId)
  if (ekf.lastTs > 0) {
    ekf.predict((now - ekf.lastTs) / 1000, input.heading)
  }
  ekf.lastTs = now

  let method: HelmetPositionMethod = 'imu_only'
  let outLat: number | null = null
  let outLng: number | null = null
  let outHeading = input.heading != null && Number.isFinite(input.heading)
    ? ((input.heading! % 360) + 360) % 360
    : ekf.headingDeg

  let gpsLat = input.lat
  let gpsLng = input.lng
  const useRelative = isPatrolRelativeGpsCamera(input.cameraId)

  if (
    useRelative
    && gpsLat != null
    && gpsLng != null
    && isValidGps(gpsLat, gpsLng)
  ) {
    ;[gpsLat, gpsLng] = mapRelativeGpsToSite(input.cameraId, gpsLat, gpsLng)
    method = 'relative'
  }

  if (gpsLat != null && gpsLng != null && isValidGps(gpsLat, gpsLng)) {
    ekf.updateGps(gpsLat, gpsLng, input.accuracyM ?? GPS_DEFAULT_ACCURACY_M)
    if (method === 'relative') method = 'relative_ekf'
    else method = 'ekf'
    const pair = ekf.latLng()
    if (pair) {
      outLat = pair[0]
      outLng = pair[1]
    }
  } else {
    const pair = ekf.latLng()
    if (pair) {
      outLat = pair[0]
      outLng = pair[1]
    }
  }

  if (outLat != null && outLng != null) {
    const [mLat, mLng] = snapPointToSite(outLat, outLng)
    outLat = mLat
    outLng = mLng
    if (method === 'relative_ekf') method = 'relative_ekf'
    else if (method === 'ekf') method = 'ekf_map'
  } else if (gpsLat != null && gpsLng != null && isValidGps(gpsLat, gpsLng)) {
    const [mLat, mLng] = snapPointToSite(gpsLat, gpsLng)
    outLat = mLat
    outLng = mLng
    method = useRelative ? 'relative' : 'map'
  }

  if (input.heading != null && Number.isFinite(input.heading)) {
    ekf.headingDeg = ((input.heading % 360) + 360) % 360
    outHeading = ekf.headingDeg
  }

  return { lat: outLat, lng: outLng, heading: outHeading, method }
}

export function mapMatchPosition(lat: number, lng: number): [number, number] {
  const [mLat, mLng] = snapPointToSite(lat, lng)
  return [mLat, mLng]
}

export function ingestHelmetImu(cameraId: string, heading: number, ts = Date.now()): void {
  const ekf = getEkf(cameraId)
  if (ekf.lastTs > 0) {
    ekf.predict((ts - ekf.lastTs) / 1000, heading)
  }
  ekf.lastTs = ts
  ekf.headingDeg = ((heading % 360) + 360) % 360
}

export function resetHelmetPositionEngine(cameraId?: string): void {
  if (!cameraId) {
    ekfByHelmet.clear()
    gpsAnchorByHelmet.clear()
    return
  }
  ekfByHelmet.delete(cameraId)
  gpsAnchorByHelmet.delete(cameraId)
}
